packages/cli/src/executions/execution.service.ts

```diff
@@ -392,67 +390,81 @@ export class ExecutionService {
 		};
 	}
 
-	/**
-	 * Stop an active execution.
-	 */
-	async stop(executionId: string) {
-		const execution = await this.executionRepository.findOneBy({ id: executionId });
+	async findAllEnqueuedExecutions() {
+		return await this.executionRepository.findMultipleExecutions(
+			{
+				select: ['id', 'mode'],
+				where: { status: 'new' },
+				order: { id: 'ASC' },
+			},
+			{ includeData: true, unflattenData: true },
+		);
+	}
+
+	async stop(executionId: string): Promise<StopResult> {
+		const execution = await this.executionRepository.findSingleExecution(executionId, {
+			includeData: true,
+			unflattenData: true,
+		});
+
+		if (!execution) throw new MissingExecutionStopError(executionId);
 
-		if (!execution) throw new NotFoundError('Execution not found');
+		this.assertStoppable(execution);
+
+		const { mode, startedAt, stoppedAt, finished, status } =
+			config.getEnv('executions.mode') === 'regular'
+				? await this.stopInRegularMode(execution)
+				: await this.stopInScalingMode(execution);
+
+		return {
+			mode,
+			startedAt: new Date(startedAt),
+			stoppedAt: stoppedAt ? new Date(stoppedAt) : undefined,
+			finished,
+			status,
+		};
+	}
 
-		if (execution.status === 'new') {
-			this.concurrencyControl.remove({ mode: execution.mode, executionId });
-			await this.executionRepository.cancel(executionId);
+	private assertStoppable(execution: IExecutionResponse) {
+		const STOPPABLE_STATUSES: ExecutionStatus[] = ['new', 'unknown', 'waiting', 'running'];
 
-			return;
+		if (!STOPPABLE_STATUSES.includes(execution.status)) {
+			throw new WorkflowOperationError(
+				`Only running or waiting executions can be stopped and ${execution.id} is currently ${execution.status}`,
+			);
 		}
+	}
 
-		const stopResult = await this.activeExecutions.stopExecution(execution.id);
+	private async stopInRegularMode(execution: IExecutionResponse) {
+		if (this.concurrencyControl.has(execution.id)) {
+			this.concurrencyControl.remove({ mode: execution.mode, executionId: execution.id });
+			return await this.executionRepository.stopBeforeRun(execution);
+		}
 
-		if (stopResult) return this.toExecutionStopResult(execution);
+		if (this.activeExecutions.has(execution.id)) {
+			await this.activeExecutions.stopExecution(execution.id);
+		}
 
-		if (this.isRegularMode) {
-			return await this.waitTracker.stopExecution(execution.id);
+		if (this.waitTracker.has(execution.id)) {
+			await this.waitTracker.stopExecution(execution.id);
 		}
 
-		// queue mode
+		return await this.executionRepository.stopDuringRun(execution);
```

Comment:
Shouldn't this be part of the responsibility of `activeExecutions` rather than being here?

Reply:
What I like about the flow right now is that it gives a clear overview of all the calls being made to stop the execution, with everything in a single spot:

```
ExecutionService.stopInRegularMode
	ConcurrencyControl.remove
	ExecutionRepository.stopBeforeRun
	// or 
	ActiveExecutions.stopExecution
	WaitTracker.stopExecution
	ExecutionRepository.stopDuringRun

ExecutionService.stopInScalingMode
	ExecutionService.stopInRegularMode
	// or
	WaitTracker.stopExecution
	this.queue.stopJob
	ExecutionRepository.stopDuringRun
```

If we moved `ExecutionRepository.stopDuringRun`
- this would require more navigation to see the full picture and 
- `ExecutionRepository.stopBeforeRun` would still remain in `ExecutionService` unless we also move that to `ConcurrencyControl.remove`

```
ExecutionService.stopInRegularMode
	ConcurrencyControl.remove
	// or 
	ActiveExecutions.stopExecution
	WaitTracker.stopExecution

ExecutionService.stopInScalingMode
	ExecutionService.stopInRegularMode
	// or
	WaitTracker.stopExecution
	this.queue.stopJob
```

The root issue in my eyes is that there is no clear distinction between what `ExecutionService` and what `ActiveExecutions` should be responsible for. Until we differentiate them better, I'd find the former flow easier to debug.